### ISSUE TITLE
False positive ExplicitCollectionElementAccessMethod on java.lang.reflect.Field.get

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#4918]
## Expected Behavior
No `ExplicitCollectionElementAccessMethod`. There's no `operator get` on `Field`.

## Observed Behavior
```
Prefer usage of the indexed access operator [] for map element access or insert methods. [ExplicitCollectionElementAccessMethod]
```

## Steps to Reproduce
```kotlin
fun reflect(f: Field) {
   val value = f.get(null) // access static field
   println(value)
}
```

## Context
I'm enabling all the checks there exists in Detekt. And I noticed this false positive. I think it's a false positive because we shouldn't encourage `[]` notation on reflection classes as they're not collections. Reflection is confusing enough already without throwing unnecessary syntactic sugar in the mix.

Note: to get ahead of conversations on exceptions to this rule: java.lang.reflect is in the standard library of Kotlin-JVM, and it is not Kotlin-friendly. The synthetic `get` from the public method is an accident.

## Your Environment

* Version of detekt used: 1.20.0
* Version of Gradle used (if applicable): 7.4.2
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Windows 10
* Link to your project (if it's a public repository): Detekt is not set up fully yet. https://github.com/TWiStErRob/net.twisterrob.sun/blob/92c3939c4f88ec0165be1bd9f4005531acfcad3e/component/paparazzi/src/main/java/net/twisterrob/sun/test/screenshots/ActivityClientSingletonHack.kt#L46-L50 is the offending file.

**Comment by TWiStErRob** [comment]
Side-track: the other way around it would make more sense. (As it is more natural indexed property access, which is chainable and more familiar to JavaScript and Groovy.)

Original code:
```kotlin
backup = mKnownInstance.get(INTERFACE_SINGLETON.get(null))
mKnownInstance.set(
	INTERFACE_SINGLETON.get(null),
	Mockito.mock(IActivityClientController::class.java)
)
```

Detekt proposed code:
```kotlin
backup = mKnownInstance[INTERFACE_SINGLETON[null]]
mKnownInstance[INTERFACE_SINGLETON[null]] = Mockito.mock(IActivityClientController::class.java)
```

Desired code:
```kotlin
backup = STATIC[INTERFACE_SINGLETON][mKnownInstance]
STATIC[INTERFACE_SINGLETON][mKnownInstance] = Mockito.mock(IActivityClientController::class.java)
```

Which is achievable, but needs these two funs:
```kotlin
const val STATIC: Any? = null

operator fun Any?.get(field: Field): Any? =
	field.get(this)

operator fun Any?.set(field: Field, value: Any?) {
	field.set(this, value)
}
```

**Comment by 3flex** [comment]
This rule used to ignore the case you've described, but the behaviour was changed in https://github.com/detekt/detekt/pull/4201.

cc @Whathecode @cortinico @BraisGabin who authored & reviewed that PR.

**Comment by Whathecode** [comment]
It's a bit unclear why you consider this a bug (using an index accessor works, right?), but I may not understand the specifics.

However, I guess the rule configuration can be extended to include configurable exceptions. Whether this specific case should be a default exception can be discussed.

**Comment by TWiStErRob** [comment]
@Whathecode I think it's two things:
 * A reflective field retrieval is in no way a "collection element access" :), so it's a false positive detection for the intention of the rule.  
   (Unless maybe you consider the JavaScript interpretation that object and `Map` is the same thing.)
 * java.lang.reflect.Field was not designed as a Kotlin API, it has no `operator get`, it just happens to have a `get` method.

**Comment by scottkennedy** [comment]
I'm seeing a similar issue on ByteBuffer:
```
    public ByteBuffer get(byte[] dst) {
        return get(dst, 0, dst.length);
    }
```

I do not think it makes sense to call `byteBuffer[destinationArray]` as a standalone call, but Detekt seems to want me to.

**Comment by BraisGabin** [comment]
With the `ByteArray` example I think that it is clear that we need to do something with this rule. The question is what we should do. I see different options:

- The name of the rule talks about collections. Maybe we could have a list of types that are collections and only force those. And let users to modify that list.
- A flag to ignore all the Java classes.
- A way to exclude some clssses/interfaces from this rule.

I vote for the second.

**Comment by TWiStErRob** [comment]
I agree with 2, with the addition that in Kotlin you can only expect `operator get` to use the indexing operator shorthand.

Essentially ignore anything that's a virtual inferred (don't remember the term) operator and only enforce where explicitly designed to do.

At this point the rule should be probably renamed to OperatorGetAsIndex rather than referring to collections.

**Comment by cortinico** [comment]
> A flag to ignore all the Java classes.

+1 for this.

**Comment by Whathecode** [comment]
> * The name of the rule talks about collections. Maybe we could have a list of types that are collections and only force those. And let users to modify that list.

This [was what was there before the PR that caused the current bug](https://github.com/detekt/detekt/pull/4201/files#diff-5bda121d91535d7dfd4b0bbdba6e83c1c28fa764ec7a51f3d715654fe62ac8bdL37) provided a more general implementation. It would also undo the desired more general behavior: [the cases where custom indexed access operators are provided](https://github.com/detekt/detekt/pull/4201/files#diff-9cdf06a75b552629026b9314753632da18b201054bac00f79959f1e2d46cbaa1R226).

Arguably, all custom [indexed access operators](https://kotlinlang.org/docs/operator-overloading.html#indexed-access-operator) relate to "collections", and the general implementation is correct: **this rule should apply.** If an API abuses indexed access operators, that would constitute a different code smell.

The original PR [did differentiate between "get" methods, and "get" methods used as indexed access operators](https://github.com/detekt/detekt/pull/4201/files#diff-9cdf06a75b552629026b9314753632da18b201054bac00f79959f1e2d46cbaa1R236). Seemingly, this does not work for some (java?) classes, which is the _actual_ underlying cause here.

> * A flag to ignore all the Java classes.

If the problem is Java classes, this may thus very well not even need to be a configurable flag. It seems the following LoC does not work as originally intended: https://github.com/detekt/detekt/pull/4201/files#diff-5bda121d91535d7dfd4b0bbdba6e83c1c28fa764ec7a51f3d715654fe62ac8bdR76

> * A way to exclude some clssses/interfaces from this rule.

I think this would be a nice to have, regardless of a fix to this bug.

**Comment by Whathecode** [comment]
While implementing a fix for this, I landed on ignoring java classes by default, but having an _opt-in_ mechanisms for known java collections.
